assets/js/custom/Translation.js

```diff
@@ -1,40 +1,134 @@
+    const commentId = $(this).attr('id').substring('comment-translation-button-'.length)
 
-  translation(
-    document.getElementById('translate-program'),
-    elementsToTranslate,
-    document.documentElement.lang
-  )
+    $(this).hide()
+
+    if (isTranslationNotAvailable(commentId)) {
+      $('#comment-translation-loading-spinner-' + commentId).show()
+      translateComment(commentId)
+    } else {
+      openTranslatedComment(commentId)
+    }
+  })
 
-  $('.comment-translate-button').each(function () {
-    const commentId = $(this).attr('id').substring('comment-translate-button-'.length)
-    translation(this, document.getElementById('comment-text-' + commentId), document.documentElement.lang)
+  $(document).on('click', '.remove-comment-translation-button', function () {
+    const commentId = $(this).attr('id').substring('remove-comment-translation-button-'.length)
+    $(this).hide()
+    $('#comment-translation-button-' + commentId).show()
+    $('#comment-translation-wrapper-' + commentId).slideUp()
+    $('#comment-text-wrapper-' + commentId).slideDown()
   })
 
-  function translation (buttonElement, textElements, targetLang) {
-    let text = ''
-    if (Array.isArray(textElements)) {
-      const array = []
-      for (let i = 0; i < textElements.length; i++) {
-        array.push(textElements[i].innerText)
+  function setTargetLanguage () {
+    targetLanguage = document.documentElement.lang
+    const decodedCookie = decodeURIComponent(document.cookie).split(';')
+    for (let i = 0; i < decodedCookie.length; i++) {
+      if (decodedCookie[i].includes('hl=') && decodedCookie[i].length < 10) {
+        targetLanguage = decodedCookie[i].substring(decodedCookie[i].indexOf('=') + 1).replace('_', '-')
       }
+    }
+  }
+
+  function getLanguageMap () {
+    $.ajax({
+      url: '../languages',
+      type: 'get',
+      success: function (data) {
+        languageMap = data
+      }
+    })
+  }
+
+  function splitTranslatedByLine () {
+    let firstLanguage = '%sourceLanguage%'
+    let secondLanguage = '%targetLanguage%'
+    if (!isSourceLanguageFirst()) {
+      firstLanguage = '%targetLanguage%'
+      secondLanguage = '%sourceLanguage%'
+    }
 
-      text = array.join('\n\n')
+    translatedByLineMap = {
+      before: translatedByLine.substring(0, translatedByLine.indexOf(firstLanguage)),
+      between: translatedByLine.substring(translatedByLine.indexOf(firstLanguage) + firstLanguage.length, translatedByLine.indexOf(secondLanguage)),
+      after: translatedByLine.substring(translatedByLine.indexOf(secondLanguage) + secondLanguage.length)
+    }
+  }
+
+  function isTranslationNotAvailable (commentId) {
+    return $('#comment-text-translation-' + commentId).attr('lang') !== targetLanguage
+  }
+
+  function isSourceLanguageFirst () {
+    return translatedByLine.indexOf('%sourceLanguage%') < translatedByLine.indexOf('%targetLanguage%')
+  }
+
+  function openGoogleTranslatePage (commentId) {

```

Comment:
I think this function can be moved to TranslateByLink file

Reply:
ie., avoid duplicating the url construction code

Reply:
So, the old translate function actually sets a href property on the button element, but I didn't want to set an href attribute since it is a button and not a link and I think if they click on the translate button again it will follow the link and open the new window. So, the openGoogleTranslatePage actually just manually opens a new window on the same page (which has the same behaviour as setting an href property on a button) and then if the user selects the button again then the API will be tried again.